xla/tests/collective_ops_e2e_test.cc

```diff
@@ -1340,6 +1341,290 @@ TEST_F(CollectiveOpsTestE2E, NoAsyncCollectives) {
+      auto dimensions = module->entry_computation()
+                            ->parameter_instruction(i)
+                            ->shape()
+                            .dimensions();
+      param_dims[i] = std::vector(dimensions.begin(), dimensions.end());
+      param_dims_per_shard[i] = param_dims[i];
+      HloSharding parameter_sharding =
+          module->entry_computation()->parameter_instruction(i)->sharding();
+      EvaluateShardedDims(param_dims_per_shard[i], param_sharded_dims[i],
+                          parameter_sharding);
+    }
+
+    // Slice the tiled inputs to match the prescribed sharding.
+    auto fake_args = xla::MakeFakeArguments(module.get()).value();
+    std::vector<std::vector<Literal>> fake_args_sliced(num_params);
+    std::vector<std::vector<Literal*>> fake_ptrs(num_partitions);
+    for (int k = 0; k < num_params; ++k) {
+      if (!param_sharded_dims[k].empty()) {
+        std::vector<int64_t> lower(param_dims_per_shard[k].size(), 0);
+        std::vector<int64_t> upper(param_dims_per_shard[k].begin(),
+                                   param_dims_per_shard[k].end());
+        for (int i = 0; i < num_partitions; ++i) {
+          fake_args_sliced[k].push_back(fake_args[k].Slice(lower, upper));
+          for (int m = param_sharded_dims.size() - 1; m >= 0; --m) {
+            if (upper[param_sharded_dims[k][m]] <
+                param_dims[k][param_sharded_dims[k][m]]) {
+              upper[param_sharded_dims[k][m]] +=
+                  param_dims_per_shard[k][param_sharded_dims[k][m]];
+              break;
+            } else {
+              upper[param_sharded_dims[k][m]] =
+                  param_dims_per_shard[k][param_sharded_dims[k][m]];
+            }
+          }
+          std::transform(upper.begin(), upper.end(),
+                         param_dims_per_shard[k].begin(), lower.begin(),
+                         std::minus<int64_t>());
+        }
+      } else {
+        fake_args_sliced[k].push_back(fake_args[k].Clone());
+      }
+    }
+    for (int k = 0; k < num_params; ++k) {
+      for (int i = 0; i < num_partitions; ++i) {
+        if (!param_sharded_dims[k].empty()) {
+          fake_ptrs[i].push_back(&fake_args_sliced[k][i]);
+        } else {
+          fake_ptrs[i].push_back(&fake_args_sliced[k][0]);
+        }
+      }
+    }
+
+    DeviceAssignment assn(/*replica_count=*/num_replicas,
+                          /*computation_count=*/num_partitions);
+    for (int64_t i = 0; i < num_partitions; ++i) {
+      assn(0, i) = i;
+    }
+    return HloTestBase::ExecuteReplicated(std::move(module), fake_ptrs,
+                                          num_partitions,
+                                          /*run_hlo_passes=*/true, &assn);
+  }
+
+  void EvaluateShardedDims(std::vector<int64_t>& dims_per_shard,
+                           std::vector<int64_t>& sharded_dims,
+                           const HloSharding& sharding) {
+    if (!sharding.IsReplicated()) {
```

Comment:
Is `IsReplicated` ever the case in this test file or can we simplify?

Reply:
Some tests use replicated shardings.